substrate/primitives/runtime/src/generic/unchecked_extrinsic.rs

```diff
@@ -33,16 +33,206 @@ use codec::{Compact, Decode, Encode, EncodeLike, Error, Input};
+/// [UncheckedExtrinsic] implementation.
+pub const EXTENSION_VERSION: ExtensionVersion = 0;
 
 /// The `SignaturePayload` of `UncheckedExtrinsic`.
-type UncheckedSignaturePayload<Address, Signature, Extra> = (Address, Signature, Extra);
+pub type UncheckedSignaturePayload<Address, Signature, Extension> = (Address, Signature, Extension);
+
+impl<Address: TypeInfo, Signature: TypeInfo, Extension: TypeInfo> SignaturePayload
+	for UncheckedSignaturePayload<Address, Signature, Extension>
+{
+	type SignatureAddress = Address;
+	type Signature = Signature;
+	type SignatureExtra = Extension;
+}
+
+/// A "header" for extrinsics leading up to the call itself. Determines the type of extrinsic and
+/// holds any necessary specialized data.
+#[derive(Eq, PartialEq, Clone)]
+pub enum Preamble<Address, Signature, Extension> {
+	/// An extrinsic without a signature or any extension. This means it's either an inherent or
+	/// an old-school "Unsigned" (we don't use that terminology any more since it's confusable with
+	/// the general transaction which is without a signature but does have an extension).
+	///
+	/// NOTE: In the future, once we remove `ValidateUnsigned`, this will only serve Inherent
+	/// extrinsics and thus can be renamed to `Inherent`.
+	Bare(ExtrinsicVersion),
+	/// An old-school transaction extrinsic which includes a signature of some hard-coded crypto.
+	Signed(Address, Signature, ExtensionVersion, Extension, ExtrinsicVersion),
+	/// A new-school transaction extrinsic which does not include a signature by default. The
+	/// origin authorization, through signatures or other means, is performed by the transaction
+	/// extension in this extrinsic.
+	General(ExtensionVersion, Extension),
+}
+
+const VERSION_MASK: u8 = 0b0011_1111;
+const TYPE_MASK: u8 = 0b1100_0000;
+const BARE_EXTRINSIC: u8 = 0b0000_0000;
+const SIGNED_EXTRINSIC: u8 = 0b1000_0000;
+const GENERAL_EXTRINSIC: u8 = 0b0100_0000;
+
+const LOWEST_SUPPORTED_EXTRINSIC_FORMAT_VERSION: ExtrinsicVersion = 4;
+
+impl<Address, Signature, Extension> Decode for Preamble<Address, Signature, Extension>
+where
+	Address: Decode,
+	Signature: Decode,
+	Extension: Decode,
+{
+	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
+		let version_and_type = input.read_byte()?;
+
+		let version = version_and_type & VERSION_MASK;
+		let xt_type = version_and_type & TYPE_MASK;
+
+		let preamble = match (version, xt_type) {
+			(
+				extrinsic_version @ LOWEST_SUPPORTED_EXTRINSIC_FORMAT_VERSION
+					..=EXTRINSIC_FORMAT_VERSION,
+				BARE_EXTRINSIC,
+			) => Self::Bare(extrinsic_version),
+			(LOWEST_SUPPORTED_EXTRINSIC_FORMAT_VERSION, SIGNED_EXTRINSIC) => {
+				let address = Address::decode(input)?;
+				let signature = Signature::decode(input)?;
+				let ext = Extension::decode(input)?;
+				Self::Signed(address, signature, 0, ext, version)
+			},
+			(EXTRINSIC_FORMAT_VERSION, SIGNED_EXTRINSIC) => {
```

Comment:
Note that I am not sure `EXTRINSIC_FORMAT_VERSION` variable makes much sense. if we bump it then it breaks the implementation of decode/encode and some other codes.
It could make more sense to hard-code V4 and V5.

Reply:
Whenever that is bumped, we have to implement new `UncheckedExtrinsic` handling logic anyway, including encoding and decoding, and we have to check all existing usage of this value.

Historically this value has been fairly stable. I think the last time this was bumped before was about 4 years ago. I'd keep as is.